api/client/client.go

```diff
@@ -95,55 +75,59 @@ func (srv *Server) AuthSign(req, ID []byte, profileName string, provider auth.Pr
 		}
 		return nil, errors.New(errors.APIClientError, errors.ServerRequestFailed)
 	}
-	result := response.Result.(map[string]interface{})
-	cert := result["certificate"].(string)
 
-	return []byte(cert), nil
+	return &response, nil
 }
 
-// Sign sends a signature request to the remote CFSSL server,
-// receiving a signed certificate or an error in response. The hostname,
-// csr, and profileName are used as with a local signing operation, and
-// the label is used to select a signing root in a multi-root CA.
-func (srv *Server) Sign(hostname string, csr []byte, profileName, label string) ([]byte, error) {
-	url := srv.getURL("sign")
-	var request = map[string]string{
-		"certificate_request": string(csr),
-		"hostname":            hostname,
-		"profile":             profileName,
-		"label":               label,
-	}
 
-	jsonData, err := json.Marshal(request)
+// AuthSign fills out an authenticated request to the server,
+// receiving a certificate or error in response.
+// It takes the serialized JSON request to send, remote address and
+// authentication provider.
+func (srv *Server) AuthSign(req, ID []byte, provider auth.Provider) ([]byte, error) {
+	url := srv.getURL("authsign")
+
+	token, err := provider.Token(req)
 	if err != nil {
-		return nil, errors.Wrap(errors.APIClientError, errors.JSONError, err)
+		return nil, errors.Wrap(errors.APIClientError, errors.AuthenticationFailure, err)
 	}
 
-	buf := bytes.NewBuffer(jsonData)
-	resp, err := http.Post(url, "application/json", buf)
-	if err != nil {
-		return nil, errors.Wrap(errors.APIClientError, errors.ClientHTTPError, err)
+	aReq := &auth.AuthenticatedRequest{
+		Timestamp:     time.Now().Unix(),
+		RemoteAddress: ID,
+		Token:         token,
+		Request:       req,
 	}
-	body, err := ioutil.ReadAll(resp.Body)
+
+	jsonData, err := json.Marshal(aReq)
 	if err != nil {
-		return nil, errors.Wrap(errors.APIClientError, errors.IOError, err)
+		return nil, errors.Wrap(errors.APIClientError, errors.JSONError, err)
 	}
-	resp.Body.Close()
 
-	var response Response
-	err = json.Unmarshal(body, &response)
+	response, err := srv.post(url, jsonData)
 	if err != nil {
-		return nil, errors.Wrap(errors.APIClientError, errors.JSONError, err)
+		return nil, err
 	}
 
-	if !response.Success || response.Result == nil {
-		if len(response.Errors) > 0 {
-			return nil, errors.Wrap(errors.APIClientError, errors.ServerRequestFailed, stderr.New(response.Errors[0].Message))
-		}
-		return nil, errors.New(errors.APIClientError, errors.ServerRequestFailed)
+	result := response.Result.(map[string]interface{})
```

Comment:
This should probably have `result, ok :=` to catch the case where the result isn't what we thought it would be.